feat: Add Dockerfile - attempt #2

richb-hanover opened the pull request:
This is a second attempt to add a Dockerfile to the PRQL project. The current state seems to work, but should be viewed as a Work in Progress. Here's what I know:

- The Dockerfile builds a container successfully pulling all its own dependencies
- Instructions for building/running are in USING_DOCKER.md
- The `Taskfile.yml` has vestiges of several tasks that were once used to develop the Docker container but are no longer in use. The 'install-brew' task installs brew with all its folderol. It is no longer used for the Docker container, but it was hard-won knowledge so I'm keeping it.
- The `Taskfile.yml` now sets up the development environment in the original manner: `cargo test` and `task setup-dev` seem to work OK on my macOS computer. I have not tested either on a clean computer.
- I have incorporated the comments from the [initial (now closed) PR](https://github.com/prql/prql/pull/1022)
- I enabled the `test-all.yml` workflow for this branch and it passed the initial round of tests
- The USING_DOCKER.md file now gives instructions for running the playground, building the book, and updating the website via the Docker container. It points readers to the various READMEs in the other component's directories.
- Does this address the concern about python and `cargo build`? (https://github.com/prql/prql/issues/1049)

richb-hanover committed:
Both files pass markdownlint

richb-hanover committed:
Merge commit '55e7c5ac4ed3af8f82ff4239e8120be2932bbe4e' into add-dockerfile

richb-hanover committed:
New tweaks for markdownlint

richb-hanover committed:
Add USING_DOCKER.md

richb-hanover committed:
Merge main branch (prql/prql) into add-dockerfile branch;
Tune up Dockerfile and Taskfile a bit;
Prepare to push to richb repository to let checks run

richb-hanover committed:
Update codeql-analysis.yaml

richb-hanover committed:
Update test-all.yaml

Add "add-dockerfile" to the push trigger

richb-hanover committed:
Merge commit '9eab4735852111ecc90341deac0952edea9ad8c4' into add-dockerfile

richb-hanover committed:
Final editorial tweaks before committing to upstream

richb-hanover committed:
Add a few words about the Developer Loop when using Docker

richb-hanover committed:
Tweak words for the Developer Loop

max-sixty commented on .github/workflows/codeql-analysis.yaml:
```diff
@@ -4,7 +4,7 @@ name: "code-ql"
 on:
   push:
-    branches: ["main"]
+    branches: ["main","add-dockerfile"]
```
```suggestion
    branches: ["main"]
```

max-sixty commented on .github/workflows/test-all.yaml:
```diff
@@ -6,6 +6,7 @@ on:
   push:
     branches:
       - main
+      - add-dockerfile
```
FYI there's a label `pr-test-all` that will do this!

richb-hanover replied:
Ahah! I'm still seeking a good github workflow (especially the Github Actions) and am not up to speed on all the "cool kid" tricks :-) 

How would I use the `pr-test-all` when I'm testing my own local branch?

max-sixty replied:
It works on PRs — so we'd need to open a PR, then add that label. I don't think it will work without a PR (though maybe you could open one locally on your fork I guess)

richb-hanover replied:
I removed it - it'll get called when my PR next hits

max-sixty commented on CHANGELOG.md:
```diff
@@ -119,69 +119,105 @@ Internal changes:
 0.2.5 is a very small release following 0.2.4 yesterday. It includes:
 
-- Add the ability to represent single brackets in an s-string, with two brackets (#752, @max-sixty )
+- Add the ability to represent single brackets in an s-string,
```
These are fine, generally would prefer in a separate PR if possible...

richb-hanover replied:
I think this  change is simply to make markdownlint happy - shortening a line that's too long.

max-sixty replied:
Hmmm, you're right. I wasn't getting these in vscode; only from the CLI. Can I ask what your setup is such that you were having these auto-format correctly?

richb-hanover replied:
I have been using `markdownlint` to detect long lines in my Markdown files, then manually breaking them at some sensible point. 

It's interesting to see VSCode and markdownlint fuss - they don't have a consistent view of the world.

max-sixty replied:
OK nice. 

FWIW I use https://marketplace.visualstudio.com/items?itemName=stkb.rewrap which is pretty good (though not always perfect, had to disable on yaml)

(But sounds like these didn't get fixed automatically; that would be ideal if `markdownlint` could do that, and we could then add to our pre-commit...)

max-sixty commented on DEVELOPMENT.md:
````diff
@@ -66,15 +73,11 @@ any inconsistencies.
 
 To build everything:
 
-```sh
````
We do want these formatted as code blocks, no?

richb-hanover replied:
Yes.

max-sixty commented on DEVELOPMENT.md:
````diff
@@ -205,24 +209,22 @@ Currently we release in a semi-automated way:
 
-  ```md
-  ## 0.2.X — [unreleased]
+## 0.2.X — [unreleased]
````
same here re code blocks?

richb-hanover replied:
Yes again.

max-sixty commented on Taskfile.yml:
```diff
@@ -72,6 +89,9 @@ tasks:
           "
+  install-precommit-install-hooks:
+    desc: pre-commit install-hooks
+      - pre-commit install-hooks
```
Does this work? It doesn't require a `cmd:`?

richb-hanover replied:
It probably didn't work - `cmds:` is present for next commit.

max-sixty commented on Taskfile.yml:
```diff
@@ -95,14 +116,43 @@ tasks:
+      - which brew
+    cmds:
+      # https://stackoverflow.com/questions/58292862/how-to-install-homebrew-on-ubuntu-inside-docker-container
+      - useradd -m -s /bin/zsh linuxbrew
```
I think we should be quite strict about not having anything like this in a Taskfile. If someone ran this by accident on their machine, and the Taskfile tried to create users, it looks suspicious / malicious, and they don't necessarily see the comment above.

Does that make sense?

We could leave them as comments in the Dockerfile? I agree it's a shame to lose them!

richb-hanover replied:
I agree. Commented out in next commit.

max-sixty commented on Taskfile.yml:
```diff
@@ -95,14 +116,43 @@ tasks:
       - sh: which brew
         msg: |
-
-
```
This is actually quite useful, otherwise it doesn't distinguish itself from the task output sufficiently (ref https://github.com/go-task/task/issues/900)

richb-hanover replied:
I see what you mean. Fixed.

max-sixty commented on Dockerfile:
```diff
@@ -0,0 +1,63 @@
+
+# ========= Install hugo =========
+RUN curl -L https://github.com/gohugoio/hugo/releases/download/v0.91.2/hugo_0.91.2_Linux-64bit.deb -o hugo.deb \
+  && apt install ./hugo.deb
```
```suggestion
  && apt install ./hugo.deb && rm hugo.deb
```
(and probably no need for it in the `.gitignore` then?

richb-hanover replied:
Yup.

max-sixty commented on Taskfile.yml:
```diff
@@ -95,14 +116,43 @@ tasks:
+      - /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
+      - chown -R $CONTAINER_USER: /home/linuxbrew/.linuxbrew
+
+  install-hugo:
```
These only work on linux so if we include them then we need a check in `preconditions` that we're on linux

richb-hanover replied:
I've commented this out and moved it to the end of the Taskfile.yml. Let's see how it looks in the next commit.

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,123 @@
+
+## Minimal testing for Dockerfile
+
+While the Dockerfile is under development, use these minimal tests
```
Having a very tight loop on PRs is a core priority for the project, so that requires having only automated tests that run before each merge – i.e. no manual tests.

At some point we could add an automated test, but until then, could we adjust this to say something like "to manually test the dockerfile"?

richb-hanover replied:
Yes.

max-sixty commented on USING_DOCKER.md:
````diff
@@ -0,0 +1,123 @@
+   ```
+   cd <directory-with-prql>
+   cargo test
+   task setup-dev
````
We already have a test for this! https://github.com/prql/prql/blob/main/.github/workflows/test-taskfile.yaml

richb-hanover replied:
Yes, but... If nothing else, it's a reminder to me about what to test before committing anything.

max-sixty commented on USING_DOCKER.md:
````diff
@@ -0,0 +1,123 @@
+
+**Language Book:** Use the command above, then enter:
+
+```
````
```suggestion
```sh
```

richb-hanover replied:
Yes

max-sixty commented on USING_DOCKER.md:
````diff
@@ -0,0 +1,123 @@
+
+```
+cd book
+mdbook serve -n 0.0.0.0 -p 3000
````
Note for the project – we could add this & the hugo command as `task`s, or add configs so we don't need to pass args to them

richb-hanover replied:
Yes - but one insurmountable hurdle at a time... :-)

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,123 @@
+# Using the Dockerfile
```
I would still vote to have this in Developing, but we can try it out here for a bit if you prefer. My logic is that people will be making a decision to develop and then thinking how they do it, rather than the other way around. And the root of the project is already quite busy, I'm planning to try and cut those files down as much as possible. (It seems like a small thing! But the more files are there, the less we can see anything about the project when viewing the GH project page!)

We should at least add a link between them if we do keep it

richb-hanover replied:
These are very real concerns from a global project standpoint. I'm quite sympathetic to your concern about a busy root directory.

At this stage, I'd counsel keeping the USING_DOCKER.md file separate. It provides a lot of motivation and hand-holding specific to the use of Docker. 

If it becomes an important tool for development, then it'll be clear how important it would be to merge the DEVELOPMENT.md and USING_DOCKER.mg files.

max-sixty replied:
Completely fine, let's see how it goes for a few months

max-sixty commented on To-do.md:
```diff
@@ -0,0 +1,3 @@
+# To Do
+
+- Cargo.toml - rust-version set to 1.63.0?
```
Is this OK?

richb-hanover replied:
The Dockerfile is tied to 1.63.0 - I think I saw someplace (in `Cargo.toml`?) that it wasn't the same Rust version...

max-sixty replied:
We pin the rust version in https://github.com/prql/prql/blob/9eab4735852111ecc90341deac0952edea9ad8c4/rust-toolchain.toml#L1-L2, and it updates automatically to be n-1 behind the latest.

I think `rustup` should automatically get the later version when that changes. We could try by setting it to `1.64.0` and ensuring it still works.

(and then at a lower cadence we could update the dockerfile so it matches and doesn't have to get a new compiler version)

richb-hanover replied:
I'm content to leave at 1.63.0 now. I'm not familiar with the rust ecosystem (and whether a version change is a "big deal" or no deal at all.) Keeping the desktop and Docker environments in sync is a Good Thing (and not much work - just update those two files, I think...)

richb-hanover committed:
Edits to USING_DOCKER.md

max-sixty reviewed (commented):
Great work @richb-hanover ! 

I added a few comments! Overall great. 

I'm a big fan of autoformatters! Which do you use? It a tad on the aggressive side! No stress apart from the areas where it makes a semantic change; e.g. a couple of places in the YAML.

Thank you! Let's merge after the next iteration, we can keep iterating later if needed but I'm excited to get this in

richb-hanover committed:
Update .github/workflows/codeql-analysis.yaml

Co-authored-by: Maximilian Roos <5635139+max-sixty@users.noreply.github.com>

richb-hanover committed:
Update markdown in DEVELOPMENT.md;
Switch to using Swatinem/rust-cachev2 globally

richb-hanover committed:
Commit

richb-hanover committed:
Add Dockerfile

richb-hanover committed:
Update USING_DOCKER.md

richb-hanover committed:
Update Dockerfile to ignore `pre-commit` since it's not necessary
   for the Docker container.
Update Taskfile, .gitignore
Bring actions/upload-artifact to v3

richb-hanover committed:
Optional pre-commit installation instructions

richb-hanover committed:
Merge branch 'main' of https://github.com/richb-hanover/prql into add-dockerfile

max-sixty commented on .github/actions/build-wheel/action.yaml:
```diff
@@ -4,7 +4,7 @@ description: "Use maturin to build python dists of prql-python. Conditional if c
   using: composite
   steps:
-    - uses: Swatinem/rust-cache@v1
+    - uses: Swatinem/rust-cache@v2
```
I think dependabot should take care of these every week or so — thanks for being proactive though!

max-sixty commented on Dockerfile:
```diff
@@ -0,0 +1,62 @@
+  && apt install ./hugo.deb \
+  && rm hugo.deb
+
+ENTRYPOINT ["/bin/bash"]
```
Overall a very nice Dockerfile — super clear, concise, thank you!

richb-hanover replied:
Thanks!

max-sixty commented on Taskfile.yml:
```diff
@@ -158,3 +178,35 @@ tasks:
+#  - the Dockerfile (installing brew takes forever)
+#    so the Dockerfile simply installs hugo & nodejs directly
+#  - the "desktop setup" - it uses other tasks in the Taskfile.yml
+# They remain in the Taskfile.yml as a hint if they should ever be needed
```
Perfect!

max-sixty commented on To-do.md:
```diff
@@ -0,0 +1,3 @@
+# To Do
+
+- Cargo.toml - rust-version set to 1.63.0?
```
I think we can remove this now, no? Fine to use for mid-PR, but eventually we should promote these to Issues

richb-hanover replied:
Yes. I wonder: should I say "- No outstanding questions" or just remove To-do.md?

richb-hanover replied:
I left it at "No outstanding ~~questions~~ issues"

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,137 @@
+When you exit the Docker container (say, at the end of the development
+session), the `prql` directory on the local machine contains the
+latest files.
+You can use `git` to pull or to push the `prql` directory as normal.
```
```suggestion
You can use `git` to pull or to push the `prql` repo as normal.
```

richb-hanover replied:
Yes

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,137 @@
+on the machine that hosts your Docker container.
+Then (one time) run `pre-commit install-hooks` to set up
+your `git` client to use the `.pre-commit-config.yaml` config file.
+These tests will run each time you commit so you can avoid silly mistakes.
```
(maybe link to this section in `DEVELOPMENT.md` in place of this paragraph?)

richb-hanover replied:
Interesting... That's another argument for separating `DEVELOPMENT.md` and `USING_DOCKER.md` - at least for the moment. That section in `DEVELOPMENT.md` tells _what_ the tests do, without telling _how_ install it (because that happens automatically with `cargo test`. 

When you're using Docker, you need to set it up separately (see the next item).

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,137 @@
+your `git` client to use the `.pre-commit-config.yaml` config file.
+These tests will run each time you commit so you can avoid silly mistakes.
+(NB: `pre-commit` must be done on the host machine
+since the Docker container only reads the repository.)
```
Is this right? How come it doesn't work in the container? Because `pre-commit` isn't installed there? (no problem tbc, no change needed)

richb-hanover replied:
Yes. All the files in the `prql` directory are controlled by tools on the host machine (git, VScode, etc.) 

The `docker run...` command "mounts" the repo "inside" the container. The container's tools then watch those directories for changes. It's very slick.

So when it comes time to commit a change, it's necessary to use `git` running on the host machine. And `pre-commit` needs to be installed on the host (not in the container).

max-sixty replied:
OK — `pre-commit`, despite its name, doesn't need to be coupled to committing. But it probably does have a git dependency, so may not work in the container. I'm not sure I get the "since the Docker container only reads the repository." line, but no great stress.

richb-hanover replied:
You are a great proofreader and make a great critique of my fuzzy thinking. I think my intuition was correct, but my explanation sloppy. Check the next commit.

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,137 @@
+
+- You'll see a `root@xxxxxxxxx:/app/#` prompt
+- Enter the commands below for the component you're working on
+- Ctl-C to exit that component
```
```suggestion
- Ctrl-C to exit that component
```

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,137 @@
+## Manual testing for Dockerfile
+
+While the Dockerfile is under development, use these minimal tests
+before committing new code.
```
Perfect, thanks!

max-sixty commented on USING_DOCKER.md:
```diff
@@ -0,0 +1,137 @@
+   then check the various components (also, as described above)
+
+1. **Examine the Github actions/workflows** for errors in
+   your own repo before pushing to the main `prql/prql` repo.
```
(No great stress, though we generally encourage early PRs, no need to look at tests on a fork before making one. I'm also not sure this bullet is docker-specific. But it's fine!)

richb-hanover replied:
OK. A difference in philosophy. I generally figure that I don't know what I'm doing (especially on a new project), and treat warnings as a cluebat.

And once the warnings are all suppressed/eliminated, I treat a new one as a bigger hint...

But once we get a few miles on this Docker thingie, then I'll feel better that it actually works properly, and we can remove [ much | most ] of this "manual test" section.

max-sixty replied:
Yes great, let's see how it goes and we'll iterate on these

max-sixty commented on .gitignore:
```diff
@@ -7,3 +7,6 @@ dist
 target
+node_modules/
+package.json
+package-lock.json
```
Surely it's quite important these two `.json` files aren't ignored? They're key files in npm projects...

richb-hanover replied:
I believe these are artifacts of installing Prettier.  At least the `package.json` contains the following. I'm not sure why they showed up so recently (as I was re-running all the installation tests last night prior to committing the most recent files.)

```
{
  "dependencies": {
    "prettier": "^2.7.1",
    "prettier-plugin-go-template": "^0.0.13"
  }
}
```

max-sixty replied:
If we can repro that, we should resolve it. But those are important files elsewhere and shouldn't be in the `.gitignore` for sure

richb-hanover replied:
Those files have never been present in the top-level prql/prql main branch. I don't know where they came from, but I'll delete them from the PR, as well as removing them from `.gitignore`. Then if they re-appear, we can be attentive...

max-sixty replied:
Great. (FYI having them in the gitignore without a `/` prefix means they're ignored everywhere, not just the top level)

max-sixty reviewed (approved):
Very nice indeed @richb-hanover !

There just two tiny changes re the `To-do.md` & `.gitignore`. And I made some very tweaky comments elsewhere, no changes necessary.

Thanks a lot!

richb-hanover committed:
Update USING_DOCKER.md

Co-authored-by: Maximilian Roos <5635139+max-sixty@users.noreply.github.com>

richb-hanover committed:
More editorial tweaks for Dockerfile instructions

richb-hanover committed:
Merge branch 'add-dockerfile' of https://github.com/richb-hanover/prql into add-dockerfile

richb-hanover committed:
Ignore node_modules, package.json, package-lock.json added by Prettier

richb-hanover committed:
Merge branch 'main' into add-dockerfile

max-sixty commented on To-do.md:
```diff
@@ -0,0 +1,3 @@
+# To Do
+
+- No outstanding issues
```
I don't think this file should be here at all — we use GitHub Issues rather than storing them in the repo. (There are lots of outstanding issues! :) )

richb-hanover replied:
I deleted it.

max-sixty replied:
~I still see it! Sorry to be a stickler for this~

richb-hanover committed:
Update USING_DOCKER.md text;
Remove To-do.md (use Github Issues instead);
